src/integration/java/io/pinecone/integration/dataplane/UpsertAndDeleteTest.java

```diff
@@ -0,0 +1,240 @@
+        // Verify the updated vector count
+        assertEquals(describeIndexStatsResponse.getNamespacesMap().get(namespace).getVectorCount(), startVectorCount - idsToDelete.length);
+    }
+
+    @Test
+    public void UpsertVectorsAndDeleteByNamespaceSyncTest() throws InterruptedException {
+        // Upsert vectors with optional parameters
+        String namespace = RandomStringBuilder.build("ns", 8);
+        UpsertResponse upsertResponse = blockingStub.upsert(buildOptionalUpsertRequest(namespace));
+
+        // Get vector count before deleting vectors
+        DescribeIndexStatsRequest describeIndexRequest = DescribeIndexStatsRequest.newBuilder().build();
+        DescribeIndexStatsResponse describeIndexStatsResponse = blockingStub.describeIndexStats(describeIndexRequest);
+        int startVectorCount = describeIndexStatsResponse.getNamespacesMap().get(namespace).getVectorCount();
+        assertEquals(startVectorCount, upsertResponse.getUpsertedCount());
+
+        // Delete all vectors in the namespace
+        DeleteRequest deleteRequest = DeleteRequest.newBuilder()
+                .setNamespace(namespace)
+                .setDeleteAll(true)
+                .build();
+        blockingStub.delete(deleteRequest);
+        Thread.sleep(3500);
+
+        // Call describeIndexStats to get updated counts
+        NamespaceSummary namespaceSummary = blockingStub.describeIndexStats(describeIndexRequest).getNamespacesMap().get(namespace);
+        // Verify the namespace is deleted
+        assertThrows(NullPointerException.class, () -> namespaceSummary.getVectorCount());
+    }
+
+    @Test
+    public void UpsertVectorsAndDeleteByFilterSyncTest() throws InterruptedException {
+        // Upsert vectors with optional and custom metadata parameters
+        String namespace = RandomStringBuilder.build("ns", 8);
+        UpsertResponse upsertResponse = blockingStub.upsert(buildOptionalUpsertRequest(namespace));
+
+        // Get vector count before deleting vectors
+        DescribeIndexStatsRequest describeIndexRequest = DescribeIndexStatsRequest.newBuilder().build();
+        DescribeIndexStatsResponse describeIndexStatsResponse = blockingStub.describeIndexStats(describeIndexRequest);
+        int startVectorCount = describeIndexStatsResponse.getNamespacesMap().get(namespace).getVectorCount();
+        assertEquals(upsertResponse.getUpsertedCount(), startVectorCount);
+
+        // Delete by filtering
+        DeleteRequest deleteRequest = DeleteRequest.newBuilder()
+                .setNamespace(namespace)
+                .setDeleteAll(false)
+                .setFilter(Struct.newBuilder()
+                        .putFields("genre", Value.newBuilder()
+                                .setStructValue(Struct.newBuilder()
+                                        .putFields("$eq", Value.newBuilder()
+                                                .setStringValue("drama")
```

Comment:
A little unfortunate that you have to have knowledge of `BuildUpsertRequest`/`buildOptionalUpsertRequest` for the metadata fields we're using to delete. I think that's fine ultimately, just something to keep in mind with helpers and where they're being used in tests.

Reply:
You're right, its a decision I made where the test is expecting the fields but this is only one instance I have seen where metadata is set, so I'll make the change now but I dont know how future patterns will look like. So a refactor might be needed in future.